### ISSUE TITLE
Replace traits with Deref

### ISSUE THREAD

**Issue by Pauan** [koute/stdweb#293]
[I had created an RFC](https://github.com/rustwasm/rfcs/pull/3) for adding in traits to wasm-bindgen (similar to what stdweb is doing), however [it has been superseded by an RFC which uses `Deref` instead](https://github.com/rustwasm/rfcs/pull/5). It is highly likely that the `Deref` RFC will win.

Although I was the biggest proponent for stdweb-style traits (and I still think it's a solid system), I've since come around to `Deref`, it has several advantages that traits don't have:

* It is much simpler to implement, and simpler to use (no need to import traits, no need to use generics).

* It doesn't cause monomorphization bloat, because it's not using generics.

* With traits you have to choose between these two styles:

   ```rust
   fn foo<A: IFoo>(x: &A) { ... }
   fn foo(x: &Foo) { ... }
   ```

   There are various trade-offs between these two styles, which makes it difficult to choose.

   But with `Deref`, you just always use `fn foo(x: &Foo) { ... }`, and it will always have the correct behavior.

* Within the documentation it lists all of the methods that the type supports (right now most of the methods are implemented within traits, so they're not obviously visible in the docs).

(There are some downsides as well, such as that users might not realize that a function/method which accepts a `Foo` also accepts a `Bar` or `Qux` (they might think it *only* accepts a `Foo`). This can largely be solved with solid documentation.)

For the sake of gaining those advantages, and also for more consistency with the wasm-bindgen ecosystem, I think we should replace our traits-based system with a `Deref`-based system.

We will need a new attribute which generates the `Deref` implementation, something like this:

```rust
#[reference(child_of = Foo)]
struct Bar( Reference );
```

That will generate this code:

```rust
impl Deref for Bar {
    type Target = Foo;

    #[inline]
    fn deref( &self ) -> &Self::Target {
        let reference: &::stdweb::Reference = value.as_ref();
        unsafe {
            <Self as ::stdweb::ReferenceType>::from_reference_unchecked_ref( reference )
        }
    }
}
```

(This also requires adding a new `unsafe fn from_reference_unchecked_ref( reference: &Reference ) -> &Self;` method on `ReferenceType`)

Top-level classes should have `#[reference(child_of = Object)]`, and `Object` should probably have `#[reference(child_of = Reference)]`  (should `Reference` have `#[reference(child_of = Value)]`?)

We will also need a migration path: first we add in `Deref` and deprecate traits, and then in the next major release we remove the traits.

I'm willing to make a PR that does all of that. But we should wait until the `Deref` RFC is accepted and implemented in wasm-bindgen.

**Comment by Pauan** [comment]
Alternatively, we could have a system where sub-classes don't contain a `Reference`, instead they contain the super-class, like this:

```rust
#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
#[reference(instance_of = "Foo")]
struct Foo( Reference );

#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
#[reference(instance_of = "Bar")]
#[reference(subclass_of(Foo))]
#[reference(child_of = Foo)]
struct Bar( Foo );

#[derive(Clone, Debug, PartialEq, Eq, ReferenceType)]
#[reference(instance_of = "Qux")]
#[reference(subclass_of(Foo, Bar))]
#[reference(child_of = Bar)]
struct Qux( Bar );
```

This makes it possible to convert from a `&Bar` to a `&Foo` without using `ReferenceType::from_reference_unchecked_ref`.

**Comment by NeverGivinUp** [comment]
I wonder, if this should be considered separately from or combined with the long term goal of making stdweb compatible to (or building stdweb upon) `wasm-bindgen`, `web-sys` and `js-sys`. @Pauan's first proposition seems to aim at the compatibility goal saying
>  for more consistency with the wasm-bindgen ecosystem

, but shys away from introducing strong dependencies like inheriting from `js-sys`' `Object` struct. @Pauan's second proposition is different to `wasm-bindgen`'s approach and does not mention the subject of compatibility (and personally I'm to ignorant to judge it myself).

**Comment by Pauan** [comment]
@NeverGivinUp It's both. The long-term goal of compatibility with wasm-bindgen will require a **lot** of changes (including breaking changes). It will take a long time and a lot of work. So obviously it must be done incrementally, in small chunks. This is one of those chunks.

But this change to `Deref` is still useful even without wasm-bindgen (due to the advantages I listed).

> shys away from introducing strong dependencies like inheriting from js-sys' Object struct. 

That's a *very* big change, requiring a dependency on wasm-bindgen and js-sys. Right now it's not even possible to combine stdweb and wasm-bindgen in the same Rust project, so that's not a feasible change right now (though it is the long-term goal).

> second proposition is different to wasm-bindgen's approach and does not mention the subject of compatibility (and personally I'm to ignorant to judge it myself).

It's not a second proposition, it's describing a small implementation detail (as an alternative to using `ReferenceType::from_reference_unchecked_ref`). The user API is the same either way (in both cases it uses `Deref`).

**Comment by NeverGivinUp** [comment]
Thanks a lot for clarifying @Pauan. Very helpful, as always (:

Obviously I didn't know what needs do be done for wasm-bindgen-compatibility and how this fits in the grand scheme of things. Is there a list of tasks?

**Comment by Pauan** [comment]
> Is there a list of tasks?

I don't think so, and it's a bit hard to make such a list, since we're not even sure what all the incompatibilities are.

As a first step, wasm-bindgen will need to support some sort of `js!` macro, [as described in this wasm-bindgen issue](https://github.com/rustwasm/team/issues/226#issuecomment-417961769).

After that it will be a matter of replacing stdweb's runtime with wasm-bindgen's runtime, replacing `Value` and `Reference` with the wasm-bindgen equivalents, replacing stdweb's ad-hoc type system with the `wasm_bindgen` macro, and then rewriting literally every single stdweb API to use the `wasm_bindgen` macro.

**Comment by koute** [comment]
Thanks for the proposal @Pauan!

Hmm... I would *probably* be fine with this, however it does make me somewhat uneasy that this wouldn't support multiple interfaces/mixins (as `Deref` can point to only one type). Are we sure this won't be a problem?

It also feels like the current trait based system is more Rust-y. Many people do think that using `Deref` to emulate inheritance is a bad practice. Even Rust's [official docs](https://doc.rust-lang.org/std/ops/trait.Deref.html) says this:

> Deref should only be implemented for smart pointers

So it's a little surprising to me that a `Deref`-based approach is likely to win.

That said, while I *probably* like trait based system more as it feels more Rust-y and is more flexible, I don't have super strong opinions that it's objectively better, although it does feel to me like it is.

You're right that if we do this it'd be the best to do this progressively. Perhaps instead of adding `#[reference(child_of=...)]` attribute we could look just inside the structure and if we see `struct Foo(Bar)` then we assume `Bar` is the parent and generate a `Deref` for it? That seems somewhat more natural and doesn't require any extra syntax.

> Obviously I didn't know what needs do be done for wasm-bindgen-compatibility and how this fits in the grand scheme of things. Is there a list of tasks?

As Pauan said, there isn't. However I would envision roughly something like this:

1) Get `wasm-bindgen` to support a basic `js_raw_asm!` macro.
2) Port our internal `__js_raw_asm!` and `js!` macros to procedural macros.
3) Make our JS macros switchable between `cargo-web`-based implementation and `wasm-bindgen` one. Now it should be possible to use `stdweb` in `wasm-bindgen` projects, although still using our own runtime.
4) Port the internals to use `js-sys` and `web-sys` and trim down our runtime. This is extra complicated by the fact that I do still want to support Emscripten for the foreseeable future. Initially this could probably be achieved by writing our own `wasm_bindgen` procedural attribute macro which would be compatible with the upstream `wasm_bindgen` macro on a source code level (we could probably reuse most of the parser and just replace the codegen part), but would generate bindings in exactly the same way as we do now. Then we could switch between `wasm-bindgen`-based runtime (which is `wasm32-unknown-unknown` only) and the current `cargo-web`-based runtime (which supports all of the targets) with a feature flag.

**Comment by koute** [comment]
> I think a good first step is to add in a `Deref`-based implementation (in addition to the current trait system), and put it under an experimental feature flag. That will let us play around with it and see how well it works, what the migration costs are, etc.

Sounds good to me!

**Comment by Pauan** [comment]
Just a heads up that the `Deref` RFC was accepted (and the traits RFC rejected), just as I predicted it would.

**Comment by derekdreery** [comment]
It's worth reading the [discussion thread for Deref RFC], it goes through quite a lot of the advantages/disadvantages of each approach.

[discussion thread for Deref RFC]: https://github.com/rustwasm/rfcs/pull/5
